File: edge_bearing/estimator.py
# edge_bearing/estimator.py

from dataclasses import dataclass, field
from typing import List, Optional, Tuple

import numpy as np

from . import framing_windowing
from . import fft
from . import gcc_phat_doa
from . import bearing_transform


@dataclass
class BearingState:
    """
    Optional internal state for smoothing across multiple chunks.
    You can extend this later (e.g., running mean, variance, last N bearings).
    """
    last_bearings: List[float] = field(default_factory=list)
    last_confidences: List[float] = field(default_factory=list)
    max_history: int = 10


def _frame_multi_channel(
    x: np.ndarray,
    fs: int,
    frame_length_ms: float = 32.0,
    hop_length_ms: float = 16.0,
    window: str = "hann",
) -> np.ndarray:
    """Frame a multi-channel signal using framing_windowing.frame_and_window.

    Args:
        x: Array of shape (n_samples, n_mics).
        fs: Sampling rate.
        frame_length_ms: Frame length in milliseconds.
        hop_length_ms: Hop length in milliseconds.
        window: Window type string passed to frame_and_window.

    Returns:
        frames: Array of shape (n_frames, frame_length, n_mics).
    """
    if x.ndim != 2:
        raise ValueError(f"Expected 2D array (n_samples, n_mics), got shape={x.shape}")

    n_samples, n_mics = x.shape
    if n_mics < 2:
        raise ValueError("Need at least 2 mics for bearing estimation")

    frames_per_ch = []
    min_frames = None

    # Frame each channel independently using the real framing_windowing API
    for ch in range(n_mics):
        frames_ch, _times = framing_windowing.frame_and_window(
            x[:, ch],
            fs,
            frame_length_ms=frame_length_ms,
            hop_length_ms=hop_length_ms,
            window_type=window,
            zero_pad=False,
            normalize_energy=False,
        )
        if min_frames is None:
            min_frames = frames_ch.shape[0]
        else:
            min_frames = min(min_frames, frames_ch.shape[0])
        frames_per_ch.append(frames_ch)

    # Trim all channels to the same number of frames and stack into 3D tensor
    assert min_frames is not None
    trimmed = [fr[:min_frames] for fr in frames_per_ch]
    frames = np.stack(trimmed, axis=-1)  # (n_frames, frame_length, n_mics)
    return frames


def _normalize_audio_shape(audio_chunk: np.ndarray) -> np.ndarray:
    """
    Ensure shape (n_samples, n_mics). Many pipelines use (n_mics, n_samples),
    so we transpose if needed.
    """
    if audio_chunk.ndim != 2:
        raise ValueError(f"audio_chunk must be 2D, got shape={audio_chunk.shape}")

    n0, n1 = audio_chunk.shape
    # heuristic: if first dim is very small (<=4) treat as (n_mics, n_samples)
    if n0 <= 4 and n1 > n0:
        return audio_chunk.T.astype(np.float32, copy=False)
    return audio_chunk.astype(np.float32, copy=False)


def _select_active_frames(frames: np.ndarray, top_k: int = 8) -> np.ndarray:
    """
    Very simple energy-based frame selection.
    frames: shape (n_frames, frame_length, n_mics)

    Returns:
        indices of frames to run DOA on.
    """
    n_frames = frames.shape[0]
    if n_frames == 0:
        return np.array([], dtype=int)

    # total energy across mics
    energy = np.mean(np.sum(frames ** 2, axis=-1), axis=1)  # (n_frames,)
    # threshold = median energy
    thr = np.median(energy)
    active = np.where(energy >= thr)[0]
    if active.size == 0:
        return np.array([], dtype=int)

    # keep at most top_k by energy
    active_sorted = active[np.argsort(energy[active])[::-1]]
    return active_sorted[: min(top_k, active_sorted.size)]


def _select_active_frames_from_spectra(spectra: np.ndarray, top_k: int = 8) -> np.ndarray:
    """
    Energy-based frame selection using frequency-domain magnitude.

    spectra: ndarray with first dimension = n_frames, remaining dims = freq / channels.
    Returns indices of frames to run DOA on.
    """
    if spectra.ndim < 2:
        return np.array([], dtype=int)

    n_frames = spectra.shape[0]
    if n_frames == 0:
        return np.array([], dtype=int)

    # Use magnitude of spectra and collapse all non-frame axes into one for energy computation.
    mag = np.abs(spectra)
    flat = mag.reshape(n_frames, -1)
    energy = np.mean(flat, axis=1)  # (n_frames,)

    thr = np.median(energy)
    active = np.where(energy >= thr)[0]
    if active.size == 0:
        return np.array([], dtype=int)

    active_sorted = active[np.argsort(energy[active])[::-1]]
    return active_sorted[: min(top_k, active_sorted.size)]


def estimate_bearing(
    audio_chunk: np.ndarray,
    fs: int,
    mic_geometry: dict,
    state: Optional[BearingState] = None,
) -> Tuple[Optional[float], float, BearingState]:
    """
    Single-node bearing estimation using:
      - framing_windowing
      - fft
      - gcc_phat_doa
      - bearing_transform

    Args:
        audio_chunk: np.ndarray of shape (n_samples, n_mics) or (n_mics, n_samples).
        fs: sampling rate (e.g., 16000 or 48000).
        mic_geometry: { "positions": ..., "orientation_deg": float, ... }
        state: previous BearingState for smoothing (can be None for stateless).

    Returns:
        bearing_deg: float in [0, 360), or None if no reliable detection.
        confidence: float in [0.0, 1.0] (rough, can be refined later).
        state_out: updated BearingState.
    """
    # 0) init state
    if state is None:
        state = BearingState()

    # 1) shape normalization: (n_samples, n_mics)
    x = _normalize_audio_shape(audio_chunk)
    n_samples, n_mics = x.shape
    if n_mics < 2:
        # if monomic, DOA is not possible (Will return None here)
        return None, 0.0, state

    # 2) framing & windowing (multi-channel)
    # Use framing_windowing.frame_and_window per channel and stack into
    # a 3D tensor of shape (n_frames, frame_length, n_mics).
    frames = _frame_multi_channel(
        x,
        fs=fs,
        frame_length_ms=32.0,
        hop_length_ms=16.0,
        window="hann",
    )

    if frames.ndim != 3:
        raise ValueError(f"Expected frames to be 3D, got shape={frames.shape}")

    n_frames, frame_length, n_mics = frames.shape

    # 3) FFT on one reference channel for simple energy-based gating.
    # fft.compute_fft_per_frame has signature
    #   compute_fft_per_frame(frames: np.ndarray, fs, nfft=None, ...)
    # and expects frames shaped (n_frames, frame_length).
    freqs, spectrum, magnitude, magnitude_db = fft.compute_fft_per_frame(
        frames[:, :, 0],  # use first mic as reference
        fs=fs,
        nfft=None,
        remove_dc=True,
        window_compensation=False,
    )

    # 4) Energy-based frame selection from frequency domain. We use
    # the magnitude spectrum as input to the selector.
    active_idx = _select_active_frames_from_spectra(magnitude, top_k=8)
    if active_idx.size == 0:
        return None, 0.0, state

    doa_angles: List[float] = []

    # Get microphone spacing and heading from geometry. These keys must
    # be provided by the caller; sensible defaults are used as fallback.
    mic_spacing_m = float(
        mic_geometry.get("mic_spacing_m",
                         mic_geometry.get("d", 0.15))  # default 15 cm
    )
    heading_deg = float(
        mic_geometry.get("heading_deg",
                         mic_geometry.get("orientation_deg", 0.0))
    )

    # Maximum TDOA based on geometry
    max_tau = mic_spacing_m / gcc_phat_doa.SOUND_SPEED

    for idx in active_idx:
        frame = frames[idx]  # (frame_length, n_mics)

        # Use first two mics for a simple linear 2-mic array model.
        sig0 = frame[:, 0]
        sig1 = frame[:, 1]

        try:
            tau = gcc_phat_doa.gcc_phat_pair(
                sig0,
                sig1,
                fs=float(fs),
                max_tau=max_tau,
                interp=16,
            )
        except Exception:
            # If GCC-PHAT fails on this frame, skip it.
            continue

        # Convert TDOA to local DoA (radians → degrees)
        theta_rad = gcc_phat_doa.tdoa_to_doa_linear(
            tau,
            d=mic_spacing_m,
            c=gcc_phat_doa.SOUND_SPEED,
        )
        theta_deg_local = float(np.degrees(theta_rad))

        # Convert local DoA to global bearing using bearing_transform API
        bearing_global = bearing_transform.local_to_global_bearing(
            theta_deg_local,
            heading_deg,
        )
        if not np.isnan(bearing_global):
            doa_angles.append(float(bearing_global))

    if not doa_angles:
        # If no DOA could be extracted, return no bearing.
        return None, 0.0, state

    # 5) Average global bearings from valid frames.
    bearing_deg_raw = float(np.mean(doa_angles))

    # 6) Simple smoothing using recent history.
    state.last_bearings.append(float(bearing_deg_raw))
    state.last_confidences.append(1.0)  # placeholder, can be refined later

    # limit history length
    if len(state.last_bearings) > state.max_history:
        state.last_bearings = state.last_bearings[-state.max_history :]
        state.last_confidences = state.last_confidences[-state.max_history :]

    # Simple moving average over recent bearings
    bearing_deg_smooth = float(np.mean(state.last_bearings))

    # confidence is the ratio of successful DOA frames to active frames (rough placeholder).
    frac = len(doa_angles) / max(1, active_idx.size)
    confidence = float(np.clip(frac, 0.0, 1.0))

    return bearing_deg_smooth, confidence, state

File: edge_webapp_adapter/src/audio/bearing/estimator.js
// src/audio/bearing/estimator.js

console.log("[Bearing] estimator.js script loading (non-module)...");

// All dependencies are expected to be attached on window.*
// from types.js, framingWindowing.js, fft.js, gccPhatDoa.js, bearingTransform.js
var frameAndWindow = window.frameAndWindow;
var computeFftPerFrame = window.computeFftPerFrame;
var gccPhatPair = window.gccPhatPair;
var tdoaToDoaLinear = window.tdoaToDoaLinear;
var SOUND_SPEED = window.SOUND_SPEED;
var localToGlobalBearing = window.localToGlobalBearing;
var createInitialBearingState = window.createInitialBearingState;

/**
 * @typedef {import("./types.js").MicGeometry} MicGeometry
 * @typedef {import("./types.js").BearingState} BearingState
 * @typedef {import("./types.js").EstimateBearingResult} EstimateBearingResult
 *
 * MultiChannelChunk: [nSamples][nMics], each sample is number.
 * @typedef {number[][]} MultiChannelChunk
 */

/**
 * Ensure shape [nSamples][nMics].
 * For now we assume audioChunk already in this shape.
 *
 * @param {MultiChannelChunk} chunk
 * @returns {MultiChannelChunk}
 */
function normalizeAudioShape(chunk) {
    if (!chunk || chunk.length === 0) return chunk;

    const nSamples = chunk.length;
    const nMics = chunk[0].length;

    // If caller gave [nMics][nSamples] style and nMics <=4, we could transpose.
    // For now, assume [nSamples][nMics].
    if (nMics <= 4) {
        return chunk;
    }

    // If needed: transpose here. Not used right now.
    return chunk;
}

/**
 * Frame a multi-channel signal: [nSamples][nMics]
 * → [nFrames][frameLength][nMics]
 *
 * @param {MultiChannelChunk} x
 * @param {number} fs
 * @param {number} [frameLengthMs=32.0]
 * @param {number} [hopLengthMs=16.0]
 * @param {"hann"|"hamming"|"blackman"|"rectangular"} [window="hann"]
 * @returns {Float32Array[][]} FrameTensor
 */
function frameMultiChannel(
    x,
    fs,
    frameLengthMs = 32.0,
    hopLengthMs = 16.0,
    window = "hann"
) {
    const nSamples = x.length;
    if (nSamples === 0) return [];

    const nMics = x[0].length;
    if (nMics < 2) {
        throw new Error("Need at least 2 mics for bearing estimation");
    }

    /** @type {Float32Array[][]} */
    const framesPerCh = [];
    let minFrames = null;

    for (let ch = 0; ch < nMics; ch++) {
        const mono = new Float32Array(nSamples);
        for (let i = 0; i < nSamples; i++) {
            mono[i] = x[i][ch];
        }

        const { frames } = frameAndWindow(mono, fs, {
            frameLengthMs,
            hopLengthMs,
            windowType: window,
            zeroPad: false,
            normalizeEnergy: false,
        });

        if (minFrames === null) minFrames = frames.length;
        else minFrames = Math.min(minFrames, frames.length);

        framesPerCh.push(frames);
    }

    if (minFrames === null || minFrames === 0) return [];

    /** @type {Float32Array[][]} */
    const frames = [];

    for (let i = 0; i < minFrames; i++) {
        const frameLength = framesPerCh[0][i].length;
        /** @type {Float32Array[]} */
        const frame = [];
        for (let j = 0; j < frameLength; j++) {
            const sample = new Float32Array(nMics);
            for (let ch = 0; ch < nMics; ch++) {
                sample[ch] = framesPerCh[ch][i][j];
            }
            frame.push(sample);
        }
        frames.push(frame);
    }

    return frames;
}

/**
 * Energy-based frame selection using magnitude spectra.
 *
 * @param {Float32Array[]} magnitude
 * @param {number} [topK=8]
 * @returns {number[]} indices
 */
function selectActiveFramesFromSpectra(magnitude, topK = 8) {
    const nFrames = magnitude.length;
    if (nFrames === 0) return [];

    const energy = new Float32Array(nFrames);
    for (let i = 0; i < nFrames; i++) {
        const mag = magnitude[i];
        let sum = 0;
        for (let k = 0; k < mag.length; k++) {
            sum += mag[k];
        }
        energy[i] = mag.length > 0 ? sum / mag.length : 0;
    }

    // median threshold
    const sorted = Array.from(energy).sort((a, b) => a - b);
    const mid = sorted[Math.floor(sorted.length / 2)];

    /** @type {number[]} */
    const active = [];
    for (let i = 0; i < nFrames; i++) {
        if (energy[i] >= mid) active.push(i);
    }
    if (active.length === 0) return [];

    // sort active by descending energy
    active.sort((a, b) => energy[b] - energy[a]);

    return active.slice(0, Math.min(topK, active.length));
}

/**
 * Single-node bearing estimation (empty DSP skeleton).
 *
 * @param {MultiChannelChunk} audioChunk - [nSamples][nMics]
 * @param {number} fs                   - sample rate (e.g. 48000)
 * @param {MicGeometry} micGeometry
 * @param {BearingState} [state]
 * @returns {EstimateBearingResult}
 */
function estimateBearing(audioChunk, fs, micGeometry, state) {
    /** @type {BearingState} */
    let s = state ?? createInitialBearingState();

    // 1) Basic input shape diagnostics
    if (!audioChunk || audioChunk.length === 0) {
        console.log("[Bearing] audioChunk empty");
        return { bearingDeg: null, confidence: 0, state: s };
    }

    const firstRow = audioChunk[0];
    const nSamplesIn = audioChunk.length;
    const nMicsIn = Array.isArray(firstRow) ? firstRow.length : "unknown";

    console.log(
        "[Bearing] audioChunk (raw) shape nSamples=",
        nSamplesIn,
        "nMics=",
        nMicsIn
    );

    const x = normalizeAudioShape(audioChunk);

    if (!x || x.length === 0 || !Array.isArray(x[0]) || x[0].length < 2) {
        console.log(
            "[Bearing] after normalize: invalid shape, nSamples=",
            x ? x.length : 0,
            "nMics=",
            x && Array.isArray(x[0]) ? x[0].length : "unknown"
        );
        return { bearingDeg: null, confidence: 0, state: s };
    }

    const nSamplesNorm = x.length;
    const nMicsNorm = x[0].length;
    console.log(
        "[Bearing] normalized shape nSamples=",
        nSamplesNorm,
        "nMics=",
        nMicsNorm
    );

    // 2) Multi-channel framing
    const frames = frameMultiChannel(
        x,
        fs,
        32.0,   // frameLengthMs
        16.0,   // hopLengthMs
        "hann" // window
    );

    const nFrames = frames ? frames.length : 0;
    console.log("[Bearing] frames.length =", nFrames);

    if (!frames || frames.length === 0) {
        console.log("[Bearing] no frames produced → returning null");
        return { bearingDeg: null, confidence: 0, state: s };
    }

    // 3) Build mono frames for FFT (reference channel = 0)
    /** @type {Float32Array[]} */
    const monoFrames = frames.map((frame, idx) => {
        const frameLength = frame.length;
        const out = new Float32Array(frameLength);
        for (let i = 0; i < frameLength; i++) {
            out[i] = frame[i][0];
        }
        return out;
    });

    const { magnitude } = computeFftPerFrame(monoFrames, fs);

    const nFramesMag = magnitude.length;
    const nBinsMag = nFramesMag > 0 && magnitude[0] ? magnitude[0].length : 0;
    console.log(
        "[Bearing] magnitude frames=",
        nFramesMag,
        "bins=",
        nBinsMag
    );

    const activeIdx = selectActiveFramesFromSpectra(magnitude, 8);
    console.log("[Bearing] activeIdx.length =", activeIdx.length);

    if (activeIdx.length === 0) {
        console.log("[Bearing] no active frames → returning null");
        return { bearingDeg: null, confidence: 0, state: s };
    }

    // Resolve mic spacing with explicit null checks (no ?? to support older iOS)
    let micSpacing = 0.15;
    if (micGeometry && micGeometry.mic_spacing_m != null) {
        micSpacing = micGeometry.mic_spacing_m;
    } else if (micGeometry && micGeometry.d != null) {
        micSpacing = micGeometry.d;
    }

    // Resolve heading (deg) with explicit null checks
    let heading = 0.0;
    if (micGeometry && micGeometry.heading_deg != null) {
        heading = micGeometry.heading_deg;
    } else if (micGeometry && micGeometry.orientation_deg != null) {
        heading = micGeometry.orientation_deg;
    }
    const maxTau = micSpacing / SOUND_SPEED;

    /** @type {number[]} */
    const bearings = [];

    for (const idx of activeIdx) {
        const frame = frames[idx];
        const frameLength = frame.length;

        const sig0 = new Float32Array(frameLength);
        const sig1 = new Float32Array(frameLength);
        for (let i = 0; i < frameLength; i++) {
            sig0[i] = frame[i][0];
            sig1[i] = frame[i][1];
        }

        let tau;
        try {
            tau = gccPhatPair(sig0, sig1, fs, maxTau, 16);
        } catch (err) {
            console.warn("[Bearing] gccPhatPair error on frame", idx, err);
            continue;
        }

        const thetaRad = tdoaToDoaLinear(tau, micSpacing, SOUND_SPEED);
        const thetaDegLocal = (thetaRad * 180) / Math.PI;
        const bearingGlobal = localToGlobalBearing(thetaDegLocal, heading);

        if (!Number.isNaN(bearingGlobal)) {
            bearings.push(bearingGlobal);
        }
    }

    console.log("[Bearing] bearings.length =", bearings.length);

    if (bearings.length === 0) {
        console.log("[Bearing] no valid bearings → returning null");
        return { bearingDeg: null, confidence: 0, state: s };
    }

    const raw =
        bearings.reduce((acc, v) => acc + v, 0) / Math.max(1, bearings.length);

    s.lastBearings.push(raw);
    s.lastConfidences.push(1.0);

    if (s.lastBearings.length > s.maxHistory) {
        s.lastBearings = s.lastBearings.slice(-s.maxHistory);
        s.lastConfidences = s.lastConfidences.slice(-s.maxHistory);
    }

    const smooth =
        s.lastBearings.reduce((acc, v) => acc + v, 0) /
        Math.max(1, s.lastBearings.length);

    const confidence = Math.min(1, bearings.length / activeIdx.length);

    console.log(
        "[Bearing] RESULT bearingDeg=",
        smooth,
        "confidence=",
        confidence
    );

    return {
        bearingDeg: smooth,
        confidence,
        state: s,
    };
}


/**
 * Small helper so app.legacy.js can be used.
 *
 * @returns {BearingState}
 */
function createBearingState() {
    return createInitialBearingState();
}


// Expose functions on window for legacy app.legacy.js
// so maybeComputeBearing() can call window.estimateBearing(...)
if (typeof window !== "undefined") {
    // attach to global for legacy script
    // eslint-disable-next-line no-undef
    window.estimateBearing = estimateBearing;
    // eslint-disable-next-line no-undef
    window.createBearingState = createBearingState;
    // mark for debugging
    // eslint-disable-next-line no-undef
    window._bearingEstimatorLoaded = true;

    console.log("[Bearing] estimator.js: attached estimateBearing to window");
}

